File: leetcode/1646.cpp
// Runtime: 0 ms, faster than 100.00% of C++ online submissions for Get Maximum in Generated Array.
// Memory Usage: 5.9 MB, less than 93.03% of C++ online submissions for Get Maximum in Generated Array.
class Solution {
 public:
  int nums_[101];
  Solution() : nums_() {
    // generate the nums
    nums_[1] = 1;
    for (int i = 1; i < 50; ++i) {
      nums_[2 * i] = nums_[i];
      nums_[2 * i + 1] = nums_[i] + nums_[i + 1];
    }

    // stash the cumulative max
    int m = INT_MIN;
    for (int i = 0; i < 101; ++i) {
      m = max(m, nums_[i]);
      nums_[i] = m;
    }
  }

  int getMaximumGenerated(int n) { return nums_[n]; }
};

File: leetcode/1646.py
"""
Runtime: 24 ms, faster than 96.48% of Python3 online submissions for Get Maximum in Generated Array.
Memory Usage: 14.3 MB, less than 47.08% of Python3 online submissions for Get Maximum in Generated Array.
"""


class T:
    def __init__(self):
        res = [0, 1] + [0] * 98
        for i in range(2, 100):
            if i % 2 == 0:
                res[i] = res[i // 2]
            elif i % 2 == 1:
                res[i] = res[i // 2] + res[i // 2 + 1]

        self.res = res


t = T()


class Solution:
    def getMaximumGenerated(self, n: int) -> int:
        return max(t.res[0 : n + 1])
